pkg/controller/controllerutil/controllerutil.go

```diff
@@ -237,10 +237,8 @@ func CreateOrPatch(ctx context.Context, c client.Client, obj client.Object, f Mu
 		if !errors.IsNotFound(err) {
 			return OperationResultNone, err
 		}
-		if f != nil {
-			if err := mutate(f, key, obj); err != nil {
-				return OperationResultNone, err
-			}
```

Comment:
Wasn't this covered by this if branch already? The mutate function wasn't being called if `f` is nil

Reply:
@vincepri in CreateOrPatch, yes, but it was not being checked in [CreateOrUpdate](https://github.com/kubernetes-sigs/controller-runtime/blob/18cb30ed465c607607dbd53dbda44ff99d265ed9/pkg/controller/controllerutil/controllerutil.go#L213), thus triggering a nil pointer dereference when mutate was nil.

Rather than adding another check in `CreateOrUpdate`, I figured it was smarter to make `mutate` handle this check as IMO callers of mutate should not have to be concerned with how to call it safely.